packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -432,8 +490,12 @@ function hydrateChildren(
     // found in this case.
     expectAddlSiblings: boolean
 ) {
-    let hasWarned = false;
+    let mismatchedChildren = false;
     let nextNode: Node | null = node;
+    const serverNodes =
+        process.env.NODE_ENV !== 'production'
+            ? Array.from(parentNode?.children).map((c) => c.cloneNode(true))
```

Comment:
```suggestion
            ? ArrayFrom(getChildNodes(parentNode), (child) => cloneNode(child, true))
```

`getChildNodes` and `cloneNode` should come from the `owner`'s `renderer`.

`parentNode` should never be nullish here, and I believe `childNodes` makes more sense since we are analyzing at the node level, not the element level, and `ShadowRoot` has `childNodes` but not `children`.

Reply:
Sounds good, thank you.

Reply:
The types in `renderer` are probably just be wrong here. This function could definitely accept a Node and work the same:

https://github.com/salesforce/lwc/blob/1a82b9941b697961aec7d05279685828eab9898e/packages/%40lwc/engine-dom/src/renderer/index.ts#L175-L177

The renderer pattern dates back to the split between engine-server and engine-dom plus some Locker integration points. Happy to discuss offline.